Colab-TextGen-GPU.ipynb

```diff
@@ -0,0 +1,144 @@
+    },
+    "accelerator": "GPU"
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/Troys-Code/AI-Automation/blob/main/Colab-TextGen-GPU.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# oobabooga/text-generation-webui\n",
+        "\n",
+        "After running both cells, a public gradio URL will appear at the bottom in a few minutes. You can optionally generate an API link.\n",
+        "\n",
+        "* Project page: https://github.com/oobabooga/text-generation-webui\n",
+        "* Gradio server status: https://status.gradio.app/"
+      ],
+      "metadata": {
+        "id": "MFQl6-FjSYtY"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#@title 1. Keep this tab alive to prevent Colab from disconnecting you { display-mode: \"form\" }\n",
+        "\n",
+        "#@markdown Press play on the music player that will appear below:\n",
+        "%%html\n",
+        "<audio src=\"https://oobabooga.github.io/silence.m4a\" controls>"
+      ],
+      "metadata": {
+        "id": "f7TVVj_z4flw"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#@title 2. Launch the web UI\n",
+        "\n",
+        "#@markdown If unsure about the branch, write \"main\" or leave it blank.\n",
+        "\n",
+        "import torch\n",
+        "from pathlib import Path\n",
+        "\n",
+        "if Path.cwd().name != 'text-generation-webui':\n",
+        "  print(\"Installing the webui...\")\n",
+        "\n",
+        "  !git clone https://github.com/oobabooga/text-generation-webui\n",
+        "  %cd text-generation-webui\n",
+        "\n",
+        "  torver = torch.__version__\n",
+        "  print(f\"TORCH: {torver}\")\n",
+        "  is_cuda118 = '+cu118' in torver  # 2.1.0+cu118\n",
+        "  is_cuda117 = '+cu117' in torver  # 2.0.1+cu117\n",
+        "\n",
+        "  textgen_requirements = open('requirements.txt').read().splitlines()\n",
+        "  if is_cuda117:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu117').replace('+cu122', '+cu117').replace('torch2.1', 'torch2.0') for req in textgen_requirements]\n",
+        "  elif is_cuda118:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu118').replace('+cu122', '+cu118') for req in textgen_requirements]\n",
+        "  with open('temp_requirements.txt', 'w') as file:\n",
+        "      file.write('\\n'.join(textgen_requirements))\n",
+        "\n",
+        "  !pip install -r extensions/openai/requirements.txt --upgrade\n",
+        "  !pip install -r temp_requirements.txt --upgrade\n",
```

Comment:
**Performance Issue**: The installation and setup process for the text-generation-webui project is inefficient and could lead to significant delays, especially when considering the repeated pip install commands and potential compatibility issues with different CUDA versions. <br> **Fix**: Optimize the setup process by creating a requirements file that dynamically adjusts to the CUDA version detected in the environment. This can be achieved by implementing a Python script that checks the CUDA version and selects the appropriate requirements file before the installation process begins. Additionally, consider using a virtual environment to isolate dependencies and prevent conflicts. <br> **Code Suggestion**: 
 ```
 -  !pip install -r extensions/openai/requirements.txt --upgrade
 -  !pip install -r temp_requirements.txt --upgrade
 +  import subprocess
 +  import re
 +  cuda_version = subprocess.check_output(["nvcc", "--version"]).decode("utf-8")
 +  match = re.search(r"release (\d+\.\d+)", cuda_version)
 +  if match:
 +      cuda = match.group(1).replace(".", "")
 +      requirements_file = f"requirements_cuda{cuda}.txt"
 +      subprocess.run(["pip", "install", "-r", requirements_file, "--upgrade"])
 ```